homeassistant/components/climate/honeywell.py

```diff
@@ -210,6 +212,24 @@ def update(self):
             self._name = data['name']
             self._is_dhw = False
 
+        if self.entity_id is None:
+            self.entity_id = async_generate_entity_id(
+                ENTITY_ID_FORMAT, self.name, hass=self.hass)
+
+        # The underlying library doesn't expose the thermostat's mode
+        # but we can pull it out of the big dictionary of information.
+
+        device_data = list(
+            filter(lambda x: x['name'] == self.name,
+                   self.device.full_data['devices']))
```

Comment:
If I'm reading the library code correctly my suggestion below should work and be a lot cleaner. We should consider renaming the instance attribute `self.device` to `self.client` on `RoundThermostat`. Having it be named `self.device` is very confusing when it's just an instance of the client, and there are other things better named devices.
```py
device = self.device.named_devices[self.name]
self.device.system_mode = device['thermostat']['changeableValues']['mode']
```